celery/datastructures.py

```diff
@@ -583,120 +583,217 @@ def values(self):
+                                         type(inserted), inserted))
+                self.add(key, inserted)
         else:
+            # AVOID THIS, it could keep old data if more parties
+            # exchange them all over and over again
             for obj in other:
                 self.add(obj)
 
+    def discard(self, item):
+        'Mark an existing item as REMOVED. If KeyError is not found, pass.'
+        entry = self._data.pop(item, self.REMOVED)
+        if entry is self.REMOVED:
+            return
+        entry[-1] = self.REMOVED
+        if self._heap_overload > self._MAX_HEAP_PERCENTS_OVERLOAD:
+            self._refresh_heap()
+
+    pop_value = discard
+
+    def purge(self, now=None):
+        """Check oldest items and remove them if needed.
+
+        :keyword now: Time of purging -- by default right now.
+                      This can be usefull for unittesting.
+        """
+        if not now:
+            now = time.time()
+        if hasattr(now, '__call__'):
+            now = now()  # if we got this now as function, evaluate it
+        if self.maxlen:
+            while len(self._data) > self.maxlen:
+                self.pop()
+        # time based expiring:
+        if self.expires:
+            while len(self._data) > self.minlen >= 0:
+                inserted_time, _ = self._heap[0]
+                if inserted_time + self.expires > now:
+                    break  # end this right now, oldest item is not expired yet
+                self.pop()
+
+    def pop(self):
+        'Remove and return the lowest time item. Return None if empty.'
+        while self._heap:
+            _, item = heappop(self._heap)
+            if item is not self.REMOVED:
+                del self._data[item]
+                return item
+        return None
+
     def as_dict(self):
-        return self._data
+        """Whole set as serializable dictionary.
+        Example::
+
+            >>> s=LimitedSet(maxlen=200)
+            >>> r=LimitedSet(maxlen=200)
+            >>> for i in range(500):
+            ...     s.add(i)
+            ...
+            >>> r.update(s.as_dict())
+            >>> r == s
+            True
+        """
+        return {key: inserted for inserted, key in self._data.values()}
 
     def __eq__(self, other):
-        return self._heap == other._heap
+        return self._data == other._data
 
     def __ne__(self, other):
         return not self.__eq__(other)
 
     def __repr__(self):
-        return 'LimitedSet({0})'.format(len(self))
+        return 'LimitedSet(maxlen={0}, expires={1}, minlen={2})' \
+            ' Current size:{3}'.format(
+                self.maxlen, self.expires, self.minlen, len(self._data))
 
     def __iter__(self):
-        return (item[1] for item in self._heap)
+        # return (item[1] for item in
+        #         self._heap if item[-1] is not self.REMOVED)
+        # ^ not ordered, slow
+        return (i for _, i in sorted(self._data.values()))
```

Comment:
well, heap is actually not ordered as you might think...
https://docs.python.org/2/library/heapq.html#priority-queue-implementation-notes says:
The interesting property of a heap is that its smallest element is always the root, heap[0].
Try it yourself: 
```py 
h=[x for x in range(10,1, -1)]
heapq.heapify(h)
print(h)
[2, 3, 4, 7, 6, 5, 8, 9, 10]
```
You might use heapq.nsmallest(2*len(h), h), which you think might be fast. But then you need to filter out removed items.   Still, sorted(self._data.values()) is as fast as heapq.nsmallest(2*len(s._heap), s._heap).  I tried to measure few things before i settled on that line :)

All the time i was coding this, while being angry at heapq, the only possible method i see to improve this code is to just use fifo queue of fixed limited size, and forget extra expiration ideas. It might make the code simpler and faster. But adding revoked items from other workers should not go to the front of the queue... it should merge in correct time -- which was not yet done correctly, by the way.